itests/kit/node_unmanaged.go

```diff
@@ -464,332 +685,424 @@ func (tm *TestUnmanagedMiner) SubmitProveCommit(
-	_ = tm.FullNode.WaitTillChain(ctx, HeightAtLeast(head.Height()+miner14.WPoStChallengeWindow+5))
+	_ = tm.FullNode.WaitTillChain(tm.ctx, HeightAtLeast(head.Height()+miner14.WPoStChallengeWindow+5))
 
 	go func() {
-		var firstPost bool
+		var postCount int
 
-		writeRespF := func(respErr error) {
-			var send WindowPostResp
-			if respErr == nil {
-				if firstPost {
-					return // already reported on our first post, no error to report, don't send anything
-				}
-				send.Posted = true
-				firstPost = true
-			} else {
-				if ctx.Err() == nil {
-					tm.t.Logf("Sector %d: WindowPoSt submission failed: %s", sectorNumber, respErr)
-				}
-				send.Error = respErr
+		recordPostOrError := func(post windowPost) {
+			head, err := tm.FullNode.ChainHead(tm.ctx)
+			if err == nil {
+				post.Epoch = head.Height()
 			}
-			select {
-			case respCh <- send:
-			case <-ctx.Done():
-			default:
+			if post.Error != nil && tm.ctx.Err() == nil {
+				tm.log("WindowPoSt submission failed for sectors %v at epoch %d: %s", post.Posted, post.Epoch, post.Error)
+			} else if tm.ctx.Err() == nil {
+				postCount++
+				tm.log("WindowPoSt loop completed post %d at epoch %d for sectors %v", postCount, post.Epoch, post.Posted)
 			}
+			tm.postsLk.Lock()
+			tm.posts = append(tm.posts, post)
+			tm.postsLk.Unlock()
 		}
 
-		var postCount int
-		for ctx.Err() == nil {
-			currentEpoch, nextPost, err := tm.calculateNextPostEpoch(ctx, sectorNumber)
-			tm.t.Logf("Activating sector %d, next post %d, current epoch %d", sectorNumber, nextPost, currentEpoch)
+		if err != nil {
+			recordPostOrError(windowPost{Error: err})
+			return
+		}
+		for tm.ctx.Err() == nil {
+			postSectors, err := tm.sectorsToPost()
 			if err != nil {
-				writeRespF(err)
+				recordPostOrError(windowPost{Error: err})
 				return
 			}
-
-			nextPost += 5 // add some padding so we're properly into the window
-
-			if nextPost > currentEpoch {
-				if _, err := tm.FullNode.WaitTillChainOrError(ctx, HeightAtLeast(nextPost)); err != nil {
-					writeRespF(err)
+			tm.log("WindowPoST sectors to post in this challenge window: %v", postSectors)
+			if len(postSectors) > 0 {
+				// something to post now
+				if err = tm.submitWindowPost(postSectors); err != nil {
+					recordPostOrError(windowPost{Error: err})
 					return
 				}
+				recordPostOrError(windowPost{Posted: postSectors})
 			}
 
-			err = tm.submitWindowPost(ctx, sectorNumber, sealedCid, sealedPath, cacheDir)
-			writeRespF(err) // send an error, or first post, or nothing if no error and this isn't the first post
-			if err != nil {
+			// skip to next challenge window
+			if err := tm.advancePostLoop(); err != nil {
+				recordPostOrError(windowPost{Error: err})
 				return
 			}
-			postCount++
-			tm.t.Logf("Sector %d: WindowPoSt #%d submitted", sectorNumber, postCount)
 		}
 	}()
+}
+
+// advancePostLoop waits until we are within the next challenge window for this miner
+func (tm *TestUnmanagedMiner) advancePostLoop() error {
```

Comment:
nit: this should be "waitForNextPostDeadline"

Reply:
👍 this function did something a bit different originally